### PR TITLE
Resnet50: add PyTorch standard case

### PR DESCRIPTION
**Pull request by yuzhou03**

### CODE REVIEW

**Comment by yuzhou03** [comment]
[nvidia-resnet50-pytorch.log.tar.gz](https://github.com/FlagOpen/FlagPerf/files/11129486/nvidia-resnet50-pytorch.log.tar.gz)

File: training/nvidia/resnet50-pytorch/README.md [hunk 1]
```diff
@@ -0,0 +1,32 @@
+### 1. 数据集准备
+[下载ImageNet2012](../../benchmarks/resnet50) 
+
+### 2. 模型Checkpoint下载
+在xxx链接下载第70个checkpoint, 并在FlagPerf/training/benchmarks/resnet50/pytorch/config/_base.py 中init_checkpoint 指定路径, 即可从第70个epoch开始继续训练。
```
Comment (upvenly) [inline training/nvidia/resnet50-pytorch/README.md:5]: ckpt 给我

File: training/benchmarks/resnet50/pytorch/config/_base.py [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+# distributed args
+
+# load and save args
+# Path to a directory containing a model checkpoint.
+init_checkpoint: str = None
+
+# data args
+# Training data dir
+data_dir: str = "/mnt/data/resnet50/train/"
```
Comment (upvenly) [inline training/benchmarks/resnet50/pytorch/config/_base.py:49]: None

File: training/run_benchmarks/config/test_conf.py [hunk 3]
```diff
@@ -28,7 +28,7 @@
 
 # The path that flagperf deploy in the cluster.
 # If not set, it will be os.path.dirname(run.py)/../../training/
-FLAGPERF_PATH_HOST = "/home/flagperf/training"
+FLAGPERF_PATH_HOST = "/home/zhouyu/workspace/FlagPerf/training"
```
Comment (upvenly) [inline training/run_benchmarks/config/test_conf.py:31]: 。。。。。基础问题勿再犯

File: training/run_benchmarks/config/cluster_conf.py [hunk 4]
```diff
@@ -1,7 +1,8 @@
 '''Cluster configs'''
 
 # Hosts to run the benchmark. Each item is an IP address or a hostname.
-HOSTS = ["10.209.20.12"]
+HOSTS = ["10.1.2.155", "10.1.2.158"]
```
Comment (upvenly) [inline training/run_benchmarks/config/cluster_conf.py:4]: 不要对外暴露IP

File: training/nvidia/resnet50-pytorch/config/config_A100x2x8.py [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+learning_rate = 0.0001
+
+beta_1: float = 0.9
+beta_2: float = 0.99
+eps: float = 1e-08
+
+seed = 23333
+max_samples_termination = 43912600
+training_event = None
+init_checkpoint = "/mnt/data/resnet50/train/checkpoint.70.pth.tar"
```
Comment (shh2000) [inline training/nvidia/resnet50-pytorch/config/config_A100x2x8.py:25]: 这个是不是最好用相对路径，benchmark/和vendor/都不使用绝对路径

File: training/nvidia/resnet50-pytorch/README.md (hunk 1 above)
Comment (yuzhou03) [reply training/nvidia/resnet50-pytorch/README.md:5]: 155机器：/raid/dataset/ImageNet/imagenet/checkpoint.70.pth.tar

File: training/nvidia/resnet50-pytorch/config/config_A100x2x8.py (hunk 5 above)
Comment (yuzhou03) [reply training/nvidia/resnet50-pytorch/config/config_A100x2x8.py]: ok

File: training/run_benchmarks/config/cluster_conf.py (hunk 4 above)
Comment (yuzhou03) [reply training/run_benchmarks/config/cluster_conf.py]: done

Comment (yuzhou03) [reply training/run_benchmarks/config/cluster_conf.py]: > 不要对外暴露IP

done

File: training/benchmarks/resnet50/pytorch/config/_base.py (hunk 2 above)
Comment (yuzhou03) [reply training/benchmarks/resnet50/pytorch/config/_base.py]: done

File: training/run_benchmarks/config/test_conf.py (hunk 3 above)
Comment (yuzhou03) [reply training/run_benchmarks/config/test_conf.py]: done

**Comment by yuzhou03** [comment]
```bash
rank0.out.log
332157:[PerfLog] {"event": "FINISHED", "value": {"e2e_time": 70834.81867527962, "global_steps": 327850, "training_images_per_second": 1185.220549865028, "converged": true, "final_loss": 1.0350726380640147, "final_acc1": 74.61336096938776, "final_acc5": 92.1273118622449, "raw_train_time": 70813.487, "init_time": 6.008, "num_trained_samples": 84557022}, "metadata": {"file": "/home/zhouyu/workspace/FlagPerf/training/benchmarks/resnet50/pytorch/run_pretraining.py", "lineno": 181, "time_ms": 1683523249922, "rank": 0}}
```

File: training/benchmarks/resnet50/pytorch/config/_base.py [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+do_train: bool = True
+
+# Number of updates steps to accumulate before performing a backward/update pass.
+gradient_accumulation_steps: int = 1
+
+# Total number of training steps to perform.
+max_steps: int = 50000000
+
+# Total number of training samples to run.
+max_samples_termination: float = 43912600
```
Comment (shh2000) [inline training/benchmarks/resnet50/pytorch/config/_base.py:69]: 这个是否应当改成epoch的形式。直接在trainer中更改为if state.epoch >= config.max_epoch或者这里的max_samples_termination=xx*epoch，xx为每个epoch的steps

File: training/benchmarks/resnet50/README.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+```
+
+## Checkpoint
+> None
+
+## AI Frameworks && Accelerators supports
+
+|            | Pytorch | Paddle | TensorFlow2 |
+| ---------- | ------- | ------ | ----------- |
+| Nvidia GPU | [✅](../../nvidia/resnet50-pytorch/README.md)       | N/A    | doing       |
```
Comment (upvenly) [inline training/benchmarks/resnet50/README.md:93]: doing？

File: training/nvidia/resnet50-pytorch/README.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
+### 1. 数据集准备
+[下载ImageNet2012](../../benchmarks/resnet50) 
+
+### 2. 模型Checkpoint下载
```
Comment (upvenly) [inline training/nvidia/resnet50-pytorch/README.md:4]: 这个应该去掉了？现在都是从头训练

File: training/nvidia/resnet50-pytorch/config/config_A100x1x2.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from config_common import *
+
+seed = 23333
+
+learning_rate = 0.1
+train_batch_size = 32
+eval_batch_size = train_batch_size
+max_steps = 10000*10000
```
Comment (upvenly) [inline training/nvidia/resnet50-pytorch/config/config_A100x1x2.py:8]: 这个值怎么来的呀

File: training/nvidia/resnet50-pytorch/config/config_A100x1x2.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+max_samples_termination = 43912600*1000
```
Comment (upvenly) [inline training/nvidia/resnet50-pytorch/config/config_A100x1x2.py:9]: 还有这个值

File: training/benchmarks/resnet50/README.md (hunk 7 above)
Comment (yuzhou03) [reply training/benchmarks/resnet50/README.md:93]: update tf2 link

File: training/benchmarks/resnet50/pytorch/config/_base.py (hunk 6 above)
Comment (yuzhou03) [reply training/benchmarks/resnet50/pytorch/config/_base.py:69]: 这里的max_steps和max_samples_termination，会被nv下的config覆盖。
已有模型的config里，这里一般也是配置的这两个参数。跑到相同的精度，epoch和原始仓库中的epoch不一定完全对得上。

File: training/nvidia/resnet50-pytorch/README.md (hunk 8 above)
Comment (yuzhou03) [reply training/nvidia/resnet50-pytorch/README.md]: removed

File: training/nvidia/resnet50-pytorch/config/config_A100x1x2.py (hunk 10 above)
Comment (yuzhou03) [reply training/nvidia/resnet50-pytorch/config/config_A100x1x2.py]: # total samples of 100 epoch for iterating ILSVRC2012 dataset 
max_samples_termination = 113116700
更新了，设置为100个epoch 遍历的样本数100*(1281167(train)+5000(validate))=133116700。

File: training/nvidia/resnet50-pytorch/config/config_A100x1x2.py (hunk 9 above)
Comment (yuzhou03) [reply training/nvidia/resnet50-pytorch/config/config_A100x1x2.py]: 更新为const了。
一般选择一个较大的数，保证在此step数之内可以达到收敛。max_steps取值的规则，perf目前还没有明确的指导建议。

File: training/benchmarks/resnet50/README.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+
+
+Please refer to this paper for a detailed description of Deep Residual network:
+[Deep Residual Learning for Image Recognition](https://arxiv.org/abs/1512.03385)
+
+
+
+
+
```
Comment (shh2000) [inline training/benchmarks/resnet50/README.md:40]: 有没有模型代码来源，就是trainer/dataloader里面那些代码？还是说resnet是咱们自己手写的各个模块
这里只放了model本身的来源

File: training/run_benchmarks/config/test_conf.py [hunk 12]
```diff
@@ -51,4 +51,5 @@
     "bert:pytorch:A100:1:8:1": "/home/datasets_ckpt/bert/train/",
     "glm:pytorch:A100:1:8:1": "/home/datasets_ckpt/glm/train/",
```
Comment (shh2000) [inline training/run_benchmarks/config/test_conf.py:52]: tc这个文件是否需要包含在commit里面？这个目录是否软链接/修正为/raid/dataset/{case}/

File: training/benchmarks/resnet50/pytorch/model/models/modeling.py [hunk 13]
```diff
@@ -0,0 +1,37 @@
+import torch
```
Comment (shh2000) [inline training/benchmarks/resnet50/pytorch/model/models/modeling.py:1]: 这个model目录额外套一个models目录的原因是什么

File: training/benchmarks/resnet50/pytorch/train/trainer_adapter.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def convert_model(model: nn.Module) -> nn.Module:
+    """convert model"""
+    return model
+
+
+def create_optimizer(model) -> Optimizer:
+    """create optimizer"""
+    optimizer = torch.optim.SGD(
```
Comment (shh2000) [inline training/benchmarks/resnet50/pytorch/train/trainer_adapter.py:17]: 优化器的具体实现是否放入适配器中（理论上优化器本身是不能改的，不能从SGD改成其他的）？是否应该按照mask-rcnn那个样例单独放到一个目录下面

File: training/benchmarks/resnet50/pytorch/train/trainer_adapter.py [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def model_to_ddp(model: nn.Module) -> nn.Module:
+    """model_to_ddp"""
+    if dist.is_available() and dist.is_initialized():
+        model = DDP(model, device_ids=[config.local_rank])
+    return model
+
+
+def create_grad_scaler():
```
Comment (shh2000) [inline training/benchmarks/resnet50/pytorch/train/trainer_adapter.py:42]: 默认不支持amp的话，可否删除这一函数

Comment (yuzhou03) [reply training/benchmarks/resnet50/pytorch/train/trainer_adapter.py:42]: removed

File: training/run_benchmarks/config/test_conf.py (hunk 12 above)
Comment (yuzhou03) [reply training/run_benchmarks/config/test_conf.py:52]: 新增一个case，test_confi.py的CASES需要添加一个 case的配置，没有问题。

File: training/benchmarks/resnet50/pytorch/train/trainer_adapter.py (hunk 14 above)
Comment (yuzhou03) [reply training/benchmarks/resnet50/pytorch/train/trainer_adapter.py]: 已移到optimizer目录

File: training/benchmarks/resnet50/pytorch/model/models/modeling.py (hunk 13 above)
Comment (yuzhou03) [reply training/benchmarks/resnet50/pytorch/model/models/modeling.py]: 已精简为model 1级目录

File: training/benchmarks/resnet50/README.md (hunk 11 above)
Comment (yuzhou03) [reply training/benchmarks/resnet50/README.md:40]: 对，后者。
resnet50的模型结构没有自行实现，用的是tv官方实现。
https://github.com/pytorch/vision/blob/main/torchvision/models/resnet.py

**Review by Ox7c000000 — APPROVED** [review]
